plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+A maximum rank of a let-bound identifier can be seen as its "deepest" (lambda,Lamdba,or let)-dependency that occurs free in that let's RHS.
+
+The depth of a lambda/Lambda-dependency that occurs free in the RHS, is calculated as: the number of lambdas that surround this lambda/Lambda + 1.
+The depth of a let-dependency that occurs free in the RHS, is the depth of the "maximum rank" calculated for that let-dependency.
+
+Any dependencies that are bound inside the rhs are simply ignored.
+
+In the special case of a letrec group, every let of that group is assigned the maximum (deepest) rank among all lets.
+
+The first-pass returns a mapping of maximum-ranks to let-identifiers, sorted on depth (see 'FloatData').
+
+-}
+
+{- Note: [Float second-pass]
+
+The second-pass first cleans the input ASTs of all let-rhs'es and puts them in a separate table 'RhsTable'.
+(actually this is yet another pass, but we don't count this cleaning pass as an individual pass, to match the pass-numbering of the original algorithm).
+
+The second pass then traverses the 'cleaned term', and whenever it reaches a lambda/Lambda position (Rank) which exists as key in the 'FloatData' table,
+it floats its corresponding lets directly inside the lambda/Lambda's body.
+The introduced lets of each Rank (level number) are not grouped in a single multi-let-rec group (as in the original algorithm), but instead in possible-multiple
+let-(nonrec & rec) groups. This groupping is done in accordance with the dependency-graph (see 'runTermDeps'):
+any let group that depends on a prior let-(rec or nonrec) group, will appear nested inside that group. The algorithm merges adjacent let-nonrec groups
+into a single let-nonrec group, since the letnonrec is linearly scoped: `(let {i=...j=...} in...) === (let i=... in let j=... in ...)`.
+
+By design choice, the second-pass preserves the floated let-bindings' original (Rec and NonRec) recursivity labels;
+it will never demote a 'Rec' let-binding to a 'NonRec', despite the demotion being valid.
+The effect is that more let-groups may be generated in a specific Rank than it would otherwise be required (optimally).
+The demotion can potentially be a nice optimization/transformation, but the way the depenndency-graph is generated for datatype binds, makes this complicated.
+See test example: `plutus-ir/test/transform/letFloat/rec3.golden`
+
+-}
+
+{- Note: [Versus original algorithm]
+
+Note that the original algorithm is applied to a different source language (Haskell/GHC), which only has 'letrec' and no support for letnonrec, type-synonym-lets, and datatype-lets.
+
+The original algorithm of that paper can be summarized to:
+
+1. The first pass of the term annotates all let-bindings with their ranks (level numbers). A rank is the maximum depth of a free-variable dependency.
+2. The second pass of the term uses those ranks "to float each binding outward to just outside the lambda which has a rank (level number) one greater than that on the binding".
```

Comment:
The description is already good, but an example here would make everything obvious.